src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs

```diff
@@ -0,0 +1,218 @@
+        private DiagnosticInfo? _lazyUseSiteDiagnostic = CSDiagnosticInfo.EmptyErrorInfo;
+
+        public static FunctionPointerMethodSymbol CreateMethodFromSource(FunctionPointerTypeSyntax syntax, Binder typeBinder, DiagnosticBag diagnostics, ConsList<TypeSymbol> basesBeingResolved, bool suppressUseSiteDiagnostics)
+        {
+            var (callingConvention, conventionIsValid) = FunctionPointerTypeSymbol.GetCallingConvention(syntax.CallingConvention.Text);
+            if (!conventionIsValid)
+            {
+                // '{0}' is not a valid calling convention for a function pointer. Valid conventions are 'cdecl', 'managed', 'thiscall', and 'stdcall'.
+                diagnostics.Add(ErrorCode.ERR_InvalidFunctionPointerCallingConvention, syntax.CallingConvention.GetLocation(), syntax.CallingConvention.Text);
+            }
+
+            RefKind refKind = RefKind.None;
+            TypeWithAnnotations returnType;
+
+            if (syntax.Parameters.Count == 0)
+            {
+                returnType = TypeWithAnnotations.Create(typeBinder.CreateErrorType());
+            }
+            else
+            {
+                var returnTypeParameter = syntax.Parameters[^1];
+
+
+                var modifiers = returnTypeParameter.Modifiers;
+                for (int i = 0; i < modifiers.Count; i++)
+                {
+                    var modifier = modifiers[i];
+                    switch (modifier.Kind())
+                    {
+                        case SyntaxKind.RefKeyword when refKind == RefKind.None:
+                            if (modifiers.Count > i + 1 && modifiers[i + 1].Kind() == SyntaxKind.ReadOnlyKeyword)
+                            {
+                                i++;
+                                refKind = RefKind.RefReadOnly;
+                            }
+                            else
+                            {
+                                refKind = RefKind.Ref;
+                            }
+
+                            break;
+
+                        case SyntaxKind.RefKeyword:
+                            Debug.Assert(refKind != RefKind.None);
+                            // A return type can only have one '{0}' modifier.
+                            diagnostics.Add(ErrorCode.ERR_DupReturnTypeMod, modifier.GetLocation(), modifier.Text);
+                            break;
+
+                        default:
+                            // '{0}' is not a valid function pointer return type modifier. Valid modifiers are 'ref' and 'ref readonly'.
+                            diagnostics.Add(ErrorCode.ERR_InvalidFuncPointerReturnTypeModifier, modifier.GetLocation(), modifier.Text);
+                            break;
+                    }
+                }
+
+                returnType = typeBinder.BindType(returnTypeParameter.Type, diagnostics, basesBeingResolved, suppressUseSiteDiagnostics);
+
+                if (returnType.IsVoidType() && refKind != RefKind.None)
```

Comment:
>if (returnType.IsVoidType() && refKind != RefKind.None) [](start = 16, length = 55)

Can we share this logic with other methods built from source?

Reply:
No, for similar reasons. Return types are parsed differently than parameter types in the main parser: if a `ref` is involved, they're parsed as a `RefTypeSyntax`. We can't reuse that in our parameter list because we need to support more than just `ref` and `ref readonly`, so we can't rely on the same paths. We _could_ potentially do this check while parsing, because having any modifiers on a `void` is never allowed, but I chose to do it here.

---
In reply to: [362940274](https://github.com/dotnet/roslyn/pull/40431#discussion_r362940274) [](ancestors = 362940274)